File: java/Algorithms/src/algorithms/QuickUnionUF.java
package algorithms;

/**
 * Created by beginor on 13-11-19.
 */
public class QuickUnionUF {

    private int[] id;

    public QuickUnionUF(int n) {
        id = new int[n];
        for (int i = 0; i < n; i++) {
            id[i] = i;
        }
    }

    private int root(int i) {
        while (i != id[i]) {
            i = id[i];
        }
        return i;
    }

    public boolean connected(int p, int q) {
        return root(p) == root(q);
    }

    public void union(int p, int q) {
        int i = root(p);
        int j = root(q);
        id[i] = j;
    }

    private void printId() {
        for (int i = 0; i < id.length; i++) {
            int item = id[i];
            System.out.print(item);
            if (i < id.length - 1) {
                System.out.print(" ");
            }
        }
        System.out.println();
    }

    public static void main(String[] args) {
        QuickUnionUF uf = new QuickUnionUF(10);
        uf.union(8, 5);
        uf.printId();
        uf.union(7, 5);
        uf.printId();
        uf.union(4, 1);
        uf.printId();
        uf.union(9, 7);
        uf.printId();
        uf.union(1, 5);
        uf.printId();
        uf.union(0, 9);
        uf.printId();
    }
}

File: dotnet/Algorithms/Algorithms/QuickUnionUF.cs
namespace Algorithms {

    public class QuickUnionUF {

        private readonly int[] id;
        private int count;

        public QuickUnionUF(int n) {
            id = new int[n];
            count = n;
            for (int i = 0; i < n; i++) {
                id[i] = i;
            }
        }

        public int Count {
            get {
                return count;
            }
        }

        public int Find(int p) {
            while (p != id[p]) {
                p = id[p];
            }
            return p;
        }

        public bool Connected(int p, int q) {
            return Find(p) == Find(q);
        }

        public void Union(int p, int q) {
            int rootP = Find(p);
            int rootQ = Find(q);
            if (rootP == rootQ) {
                return;
            }
            id[rootP] = rootQ;
            count--;
        }
    }
}
